### PR TITLE
#36 add sync scheduler

### PR DESCRIPTION
**Pull request by wlthomson**
Adds basic sync scheduling for periodically pushing sensor data to upstream server.

### CODE REVIEW

File: src/features/Sync/SyncSlice.ts [hunk 1]
```diff
@@ -398,15 +478,65 @@ const getPassword = (state: RootState): string => {
...
+    const syncOutManager = yield call(DependencyLocator.get, DEPENDENCY.SYNC_OUT_MANAGER);
+    try {
+        const loginUrl = yield call(syncOutManager.getLoginUrl);
+        const sensorUrl = yield call(syncOutManager.getSensorUrl);
+        const temperatureLogUrl = yield call(syncOutManager.getTemperatureLogUrl);
+        const temperatureBreachUrl = yield call(syncOutManager.getTemperatureBreachUrl);
+        const username = yield call(syncOutManager.getUsername);
+        const password = yield call(syncOutManager.getPassword);
+        const lastSync = yield call(syncOutManager.getLastSync);
+        const isPassiveSyncEnabled = yield call(syncOutManager.getIsPassiveSyncEnabled)
```
Comment (joshxg) [inline src/features/Sync/SyncSlice.ts:519]: I think adding a single facade func `getSyncConfig` would be good here

File: src/features/Sync/SyncSlice.ts [hunk 2]
```diff
@@ -544,51 +710,55 @@ function* syncTemperatureBreaches({ payload: { temperatureBreachUrl }}: SyncTemp
...
+}
+
+function* startSyncScheduler(): SagaIterator {
+    while (true) {
+        const loginUrl = yield select(getLoginUrl);
+        const sensorUrl = yield select(getSensorUrl);
+        const temperatureLogUrl = yield select(getTemperatureLogUrl);
+        const temperatureBreachUrl = yield select(getTemperatureBreachUrl);
+        const username = yield select(getUsername);
+        const password = yield select(getPassword);
```
Comment (joshxg) [inline src/features/Sync/SyncSlice.ts:734]: I think avoiding use of the store when you can is good in the sagas. The saga is now dependent on when the store is rehydrated, which is dependent on the implementation in the UI and the database. 

If we were to fetch these details from the db (with a `getConfig` method 🤣 ), we are only dependent on the db

File: src/ui/screens/settings/SyncSettingsScreen.tsx [hunk 3]
```diff
@@ -84,6 +91,21 @@ export const SyncSettingsScreen: FC = () => {
           subtext="Syncs all records currently stored in sync queue"
           onPress={() => dispatch(SyncAction.syncAll(loginUrl, sensorUrl, temperatureLogUrl, temperatureBreachUrl, username, password))} 
         />
+        {
+          isPassiveSyncEnabled ? (
+            <SettingsButtonRow
+              label="Stop sync scheduler"
+              subtext={`Stops sync scheduler (last sync: ${moment.unix(lastSync).format(FORMAT.DATE.STANDARD_DATE)})`}
```
Comment (joshxg) [inline src/ui/screens/settings/SyncSettingsScreen.tsx:98]: I think we should (in the future?) not import `moment` and instead use `formatService` to do all the formatting of dates etc.

moment is essentially getting deprecated and is a good reminder that any dependency might need to be swapped out!

File: src/features/Sync/SyncSlice.ts [hunk 4]
```diff
@@ -544,51 +710,55 @@ function* syncTemperatureBreaches({ payload: { temperatureBreachUrl }}: SyncTemp
...
-            password,
-        };
-
-        yield put(SyncAction.fetchAllSuccess(syncConfig));
-    } catch (error) {
-        // eslint-disable-next-line no-console
-        console.log(error.message);
-        // TODO: add logic for handling failed pushes
+    const authenticateResult = yield take([SyncAction.authenticateSuccess, SyncAction.authenticateFailure]);
+    if (authenticateResult.type === SyncAction.authenticateSuccess.type) {
```
Comment (joshxg) [inline src/features/Sync/SyncSlice.ts:716]: could possibly only take the success and instead spawn a watcher for failure which does the reset .. could be nice.. could be overkill 🤔

**Review by joshxg — APPROVED** [review]
Nice. No show stoppers just suggestions for what I think is easier to grok as a reader of the code!

File: src/features/Sync/SyncSlice.ts (hunk 2 above)
Comment (wlthomson) [reply src/features/Sync/SyncSlice.ts:734]: Resolved by #91.

File: src/ui/screens/settings/SyncSettingsScreen.tsx (hunk 3 above)
Comment (wlthomson) [reply src/ui/screens/settings/SyncSettingsScreen.tsx:98]: Issue created (#85).

**Pull request merged**